#VRML_SIM R2025a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/RectangleArena.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/appearances/protos/MattePaint.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/appearances/protos/Rubber.proto"
EXTERNPROTO "../protos/Roadmark.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/appearances/protos/Parquetry.proto"

WorldInfo {
  info [
    "Model Vehicle (NXP Cup Alamak) Digital Twin"
  ]
  title "Alamak Model Vehicle"
  basicTimeStep 8
  FPS 45
  contactProperties [
    ContactProperties {
      material1 "wheel_rubber"
    }
  ]
}
Viewpoint {
  orientation 0.13728662992660678 0.7782552679641417 -0.612756982114 0.5605888048316012
  position -2.7199098888287767 0.5872278990286703 1.8337513105998757
  follow "Alamak"
  followType "Mounted Shot"
}
TexturedBackground {
}
TexturedBackgroundLight {
}
RectangleArena {
  floorSize 50 50
  floorTileSize 2 2
  floorAppearance Parquetry {
    type "light strip"
    colorOverride 0.1 0.1 0.1
  }
}
DEF ALAMAK Robot {
  translation 0 0 0.031
  children [
    DEF ALAMAK_BODY_SH Shape {
      appearance MattePaint {
        baseColor 0.168627 0.168627 0.168627
      }
      geometry Box {
        size 0.265 0.125 0.025
      }
    }
    DEF ALAMAK_RL_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.12 0.08 0
        dampingConstant 0.001
        staticFriction 0.002
      }
      device [
        PositionSensor {
          name "rl_encoder"
          resolution 0.00153398
        }
        RotationalMotor {
          name "rl_motor"
          controlPID 1 0 0.1
          maxVelocity 285
          maxTorque 0.2294
        }
      ]
      endPoint Solid {
        translation -0.12000064696667458 0.08000075424238581 -5.762730201246555e-09
        rotation 0 1 0 0
        children [
          DEF ALAMAK_WHEEL_POSE Pose {
            rotation 1 0 0 -1.5707953071795862
            children [
              DEF ALAMAK_WHEEL_SH Shape {
                appearance Rubber {
                }
                geometry DEF ALAMAK_WHEEL_CYL Cylinder {
                  height 0.026
                  radius 0.032
                  subdivision 64
                }
              }
            ]
          }
        ]
        name "alamak_rl_wheel"
        contactMaterial "wheel_rubber"
        boundingObject USE ALAMAK_WHEEL_POSE
        physics DEF ALAMAK_WHEEL_PHYSICS Physics {
          density -1
          mass 0.05
        }
      }
    }
    DEF ALAMAK_RR_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.12 -0.08 0
        dampingConstant 0.001
        staticFriction 0.002
      }
      device [
        RotationalMotor {
          name "rr_motor"
          controlPID 1 0 0.1
          maxVelocity 285
          maxTorque 0.2294
        }
        PositionSensor {
          name "rr_encoder"
          resolution 0.00153398
        }
      ]
      endPoint Solid {
        translation -0.12000144537206893 -0.08000159518239915 -1.4540245600688463e-06
        rotation 0 1 0 0
        children [
          USE ALAMAK_WHEEL_POSE
        ]
        name "alamak_rr_wheel"
        contactMaterial "wheel_rubber"
        boundingObject USE ALAMAK_WHEEL_POSE
        physics USE ALAMAK_WHEEL_PHYSICS
      }
    }
    DEF ALAMAK_FL_WHEEL Hinge2Joint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor 0.065 0.08 0
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      device [
        RotationalMotor {
          name "fl_servo"
          maxVelocity 7.48
          minPosition -0.45
          maxPosition 0.45
          maxTorque 1.08
        }
        PositionSensor {
          name "fl_servo_encoder"
        }
      ]
      endPoint Solid {
        translation 0.06500092337164304 0.08000155513982896 9.557792592957768e-07
        children [
          USE ALAMAK_WHEEL_POSE
        ]
        name "alamak_fl_wheel"
        contactMaterial "wheel_rubber"
        boundingObject USE ALAMAK_WHEEL_POSE
        physics USE ALAMAK_WHEEL_PHYSICS
      }
    }
    DEF ALAMAK_FR_WHEEL Hinge2Joint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor 0.065 -0.08 0
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      device [
        RotationalMotor {
          name "fr_servo"
          maxVelocity 7.48
          minPosition -0.45
          maxPosition 0.45
          maxTorque 1.08
        }
        PositionSensor {
          name "fr_servo_encoder"
        }
      ]
      endPoint Solid {
        translation 0.06499964028414816 -0.0800008969279732 1.1203611898664173e-06
        children [
          USE ALAMAK_WHEEL_POSE
        ]
        name "alamak_fr_wheel"
        contactMaterial "wheel_rubber"
        boundingObject USE ALAMAK_WHEEL_POSE
        physics USE ALAMAK_WHEEL_PHYSICS
      }
    }
    Gyro {
      translation -0.0275 0 0
      name "imu_gyro"
    }
    InertialUnit {
      translation -0.03 0 0.05
      name "imu_inertial_unit"
    }
    Accelerometer {
      translation -0.03 0 0.05
      name "imu_accel"
      resolution 0.0383
    }
    Compass {
      translation -0.03 0 0.05
      name "imu_compass"
    }
    Camera {
      translation 0.065 0 0.125
      rotation 0 1 0 0
      children [
        Solid {
          children [
            Pose {
              rotation 1 0 0 -1.5708
              children [
                DEF BOARD Group {
                  children [
                    Shape {
                      appearance PBRAppearance {
                        baseColorMap ImageTexture {
                          url [
                            "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/robots/nvidia/jetbot/protos/textures/raspberry_pi_camera.jpg"
                          ]
                        }
                        roughness 1
                        metalness 0
                      }
                      geometry IndexedFaceSet {
                        coord Coordinate {
                          point [
                            -0.0005 -0.01199995592153876 -0.011999999999919044
                            -0.0005 0.01200004407846124 -0.011999999999919044
                            -0.0005 -0.01200004407846124 0.011999999999919044
                            -0.0005 0.01199995592153876 0.011999999999919044
                            0.0005 -0.01199995592153876 -0.011999999999919044
                            0.0005 0.01200004407846124 -0.011999999999919044
                            0.0005 -0.01200004407846124 0.011999999999919044
                            0.0005 0.01199995592153876 0.011999999999919044
                          ]
                        }
                        texCoord TextureCoordinate {
                          point [
                            0 0
                            1 0
                            0 1
                            1 1
                            0.0001 0.0001
                            0.9999 0.0001
                            0.0001 0.9999
                            0.9999 0.9999
                          ]
                        }
                        coordIndex [
                          1, 0, 2, 3, -1, 4, 5, 7, 6, -1, 5, 1, 3, 7, -1, 0, 4, 6, 2, -1, 7, 3, 2, 6, -1, 4, 0, 1, 5, -1
                        ]
                        texCoordIndex [
                          1, 0, 2, 3, -1, 0, 1, 3, 2, -1, 1, 5, 7, 3, -1, 0, 4, 6, 2, -1, 7, 3, 2, 6, -1, 0, 4, 5, 1, -1
                        ]
                      }
                    }
                  ]
                }
              ]
            }
            Pose {
              translation -0.003 0 -0.0015
              children [
                Shape {
                  appearance DEF LIGHT_GRAY_APPEARANCE PBRAppearance {
                    baseColor 0.729412 0.741176 0.713725
                    roughness 0.5
                    metalness 0.2
                  }
                  geometry Box {
                    size 0.004 0.022 0.027
                  }
                }
              ]
            }
          ]
        }
      ]
      name "rpi_camera_v2"
      fieldOfView 1.0472
      width 820
      height 616
    }
  ]
  name "Alamak"
  description "NXP Cup Alamak - model vehicle"
  boundingObject USE ALAMAK_BODY_SH
  physics Physics {
    density -1
    mass 4
    centerOfMass [
      -0.0275 0 0
    ]
  }
  controller "alamak"
  supervisor TRUE
}
DEF TRACK Group {
  children [
    DEF RM_S1_1 Roadmark {
      translation 0.1 -0.1 0
      name "rm_s1_1"
    }
    DEF RM_S1_2 Roadmark {
      translation 0.5 0 0
      name "rm_s1_2"
    }
    DEF RM_S1_3 Roadmark {
      translation 1 0 0
      name "rm_s1_3"
    }
    DEF RM_S1_4 Roadmark {
      translation 1.5 0 0
      name "rm_s1_4"
    }
    DEF RM_S1_4 Roadmark {
      translation 1.9 -0.05 0
      name "rm_s1_4(2)"
    }
    DEF RM_S1_4 Roadmark {
      translation 1.5 0 0
      name "rm_s1_4(1)"
    }
    DEF RM_C1_1 Roadmark {
      translation 2.2 -0.2 0
      name "rm_c1_1"
    }
    DEF RM_C1_2 Roadmark {
      translation 2.5 -0.5 0
      name "rm_c1_2"
    }
    DEF RM_C1_3 Roadmark {
      translation 2.7 -0.8 0
      name "rm_c1_3"
    }
    DEF RM_C1_4 Roadmark {
      translation 2.9 -1.1 0
      name "rm_c1_5"
    }
    DEF RM_S2_1 Roadmark {
      translation 3.1 -1.3 0
      name "rm_s2_1"
    }
    DEF RM_S2_1 Roadmark {
      translation 3.4 -1.5 0
      name "rm_s2_1(1)"
    }
    DEF RM_S2_2 Roadmark {
      translation 4.6 -1.5 0
      name "rm_s2_2"
    }
    DEF RM_S2_2 Roadmark {
      translation 4.3 -1.5 0
      name "rm_s2_2(1)"
    }
    DEF RM_S2_2 Roadmark {
      translation 4 -1.5 0
      name "rm_s2_2(2)"
    }
    DEF RM_S2_2 Roadmark {
      translation 3.7 -1.5 0
      name "rm_s2_2(3)"
    }
    DEF RM_S2_3 Roadmark {
      translation 4.9 -1.5 0
      name "rm_s2_3"
    }
    DEF RM_C2_1 Roadmark {
      translation 0.2 -2.2 -1
      name "rm_c2_1"
    }
    DEF RM_S3_1 Roadmark {
      translation 5.2 -1.5 0
      name "rm_s3_1"
    }
    DEF RM_S3_11 Roadmark {
      translation 5.7 -1.5 0
      name "rm_s3_1(1)"
    }
    DEF RM_S3_12 Roadmark {
      translation 6.2 -1.5 0
      name "rm_s3_1(2)"
    }
    DEF RM_C3_6 Roadmark {
      translation 0 -1.5 -1
      name "rm_c3_6"
    }
    DEF RM_C3_6 Roadmark {
      translation 0.05 -1.9 -1
      name "rm_c3_6(1)"
    }
    DEF RM_S4_1 Roadmark {
      translation 0 -1 -1
      name "rm_s4_1"
    }
    DEF RM_S4_2 Roadmark {
      translation 0 -0.5 -1
      name "rm_s4_2"
    }
  ]
}
